[fix bug 1301721] isFirefoxUpToDate should only use the major version until FF 49 is released

alexgibson opened the pull request:
## Description
- Switches checks for `_isFirefoxUpToDate()` to only use the major version number until Firefox 49 is released (see bug for context).

## Bugzilla link
https://bugzilla.mozilla.org/show_bug.cgi?id=1301721

## Testing
- Check in both current & older Firefox (or change UA) to ensure no regressions.

kyoshino commented:
Yep, that's the trick used in [new/scene1.js](https://github.com/mozilla/bedrock/blob/master/media/js/firefox/new/scene1.js#L54-L56) 👍

jpetto commented:
All changes here look good. Tested in latest Fx and with overridden UA string and all is as expected.

Should we also remove the [`data.isUpToDate` check on scene 1](https://github.com/mozilla/bedrock/blob/master/media/js/firefox/new/scene1.js#L63) of `/firefox/new/`? If I understand the bug correctly, the code as-is would result in Fx release users on OS X or Linux incorrectly not being offered the "Refresh" button and survey.

FWIW, looking at the URLs affected in this PR on production, Fx 48.0 on OS X is treated as up to date. Is this expected?

⌛ 

Hm, after following the ridiculous path to get to a Firefox release download while using Firefox 👿 , I see 48.0.2 is now available for OS X (even though while using 48.0 I was not offered an update in the about screen). Could something have changed on the product details/release end that would make these changes unnecessary? (I definitely have not tested on Linux.)

jpetto commented:
Downloaded 48.0.1 and I was able to reproduce the bug. I'm sure there's a reason 48.0.1 is affected but 48.0 is not...

Also, about screen in 48.0.1 says it's up to date. 😖

alexgibson commented:
48.0.2 is available for download but automatic updates are throttled, so this change is still required.

Strange re: 48.0 behaving differently. Not sure what's up there will require some debugging in the morning.

jpetto commented:
If it helps the debugging, adrian in #webdev tested in 48.0 on Ubuntu and experienced the same behavior as I.

alexgibson commented:
I don't think the platform should make much difference but thanks for double checking. I think there's a bug in our client code, will see if I can track it down and fix while we're here.

alexgibson commented:
@jpetto - thanks for the review, I updated `scene.js` to also use `client._isFirefoxUpToDate(false)`

I also tracked down the bug in our client code (looks like due to a typo in one of our unit tests 😢). I fixed the typo and updated the logic so the tests all pass. Testing in 48.0 (against 48.0.1 as latest) it now seems to work as expected. I'd appreciate some extra review here, thanks.

jpetto commented on media/js/base/mozilla-client.js:
```diff
@@ -158,16 +158,32 @@ if (typeof Mozilla === 'undefined') {
                     break;
+                } if (userVerNumber > latestVerNumber) {
+                    isUpToDate = true;
+                    break;
```
The new logic here is sound as far as I can tell/test. However, it's really hard to follow this function - the complexity has exceeded reasonable variable names. If you have the time (and freshness of mind), it would be great to add some detailed comments.

alexgibson replied:
Yeah this function is a bit overly long it does seem complicated. I think I may actually split the version check here out into a separate function.

jpetto commented on tests/unit/spec/base/mozilla-client.js:
```diff
@@ -458,7 +458,7 @@ describe('mozilla-client.js', function() {
         it('should consider outdated if user version is slightly less than latest version and strict option is true', function() {
             test(true, false, '46.0.1').toBeFalsy();
-            test(true, false, '45.0').toBeFalsy();
+            test(true, false, '46.0').toBeFalsy();
```
Maybe add a comment here stating that it's important to test the scenario of user version 46.0 against latest version 46.0.1. (I couldn't come up with a non-awkward phrase to describe that general test case. 😕)

alexgibson replied:
I think the test description here sums already sums it up about as elequently as another comment could accomplish?

jpetto replied:
Agreed. 👍

alexgibson commented on media/js/firefox/new/scene1.js:
```diff
@@ -60,7 +60,8 @@
                     // data.accurate will only be true if UITour API is working.
-                    if (data.channel === 'release' && data.isUpToDate && data.accurate) {
+                    // bug 1301721 only use major Firefox version until 49.0 is released
+                    if (data.channel === 'release' && client._isFirefoxUpToDate(false) && data.accurate) {
```
I think calling it again here explicitly is fine, as it also serves as a reminder that we need to change it back again in the future.

alexgibson replied:
On second thought, I will remove this bug just leave the comment 👍

alexgibson replied:
Come to think of it, we probably don't need the additional check in here at all (showing it for the major release version number is consistent with the rest of the page logic). I'll just remove it.

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -158,16 +158,32 @@ if (typeof Mozilla === 'undefined') {
+                if (userVerNumber < latestVerNumber) {
                     isUpToDate = false;
                     break;
+                } if (userVerNumber > latestVerNumber) {
```
Bah, good catch

alexgibson commented:
@jpetto - split out the two functions and added comments, so hopefully the logic is a bit easier to follow. Tests still passing 👍

jpetto commented:
I'd say this one is good to go. r+

Nice (grueling) work! 😩
